crates/sui-graphql-rpc/src/server/sui_sdk_data_provider.rs

```diff
@@ -175,6 +181,65 @@ impl DataProvider for SuiClient {
         Ok(connection)
     }
 
+    async fn fetch_checkpoint_connection(
+        &self,
+        first: Option<u64>,
+        after: Option<String>,
+        last: Option<u64>,
+        before: Option<String>,
+    ) -> Result<Connection<String, Checkpoint>> {
+        ensure_forward_pagination(&first, &after, &last, &before)?;
+
+        let count = first.map(|q| q as usize);
+
+        let after_parsed = if let Some(after) = after.map(|x| x.parse::<u64>()) {
+            match after {
+                Ok(after_u64) => Some(SerdeBigInt::from(after_u64)),
+                _ => {
+                    return Err(Error::InvalidCursor(
+                        "Cannot convert after parameter into u64 in the checkpoint connection"
+                            .to_string(),
+                    ))
+                    .extend()
+                }
+            }
+        } else {
+            None
+        };
+
+        let pg = self
+            .read_api()
+            .get_checkpoints(after_parsed, count, false)
+            .await?;
+
+        let system_state = self.governance_api().get_latest_sui_system_state().await?;
+        let protocol_configs = self.fetch_protocol_config(None).await?;
+
+        let data = pg
+            .data
+            .iter()
+            .map(|c| convert_json_rpc_checkpoint(c, &system_state, &protocol_configs))
+            .collect::<Vec<_>>();
+
+        let checkpoints = if data.iter().any(|x| x.is_err()) {
+            return Err(Error::Internal(
+                "Cannot convert the JSON RPC checkpoint into GraphQL checkpoint type".to_string(),
+            ))
+            .extend();
+        } else {
+            data.into_iter().map(|x| x.unwrap()).collect::<Vec<_>>()
+        };
```

Comment:
You can `collect` an iterator where the item is `Result<T, _>` into a `Result<Vec<T>, _>` instead of a `Vec<Result<T>, _>` and it will either produce the first error it could find or all the `Ok(_)` data.

```suggestion
        let data: Result<Vec<_>, _> = pg
            .data
            .iter()
            .map(|c| convert_json_rpc_checkpoint(c, &system_state, &protocol_configs))
            .collect();
            
        let checkpoints = data.map_err(|e| Error::Internal(
            format!("Cannot convert the JSON RPC checkpoint into GraphQL checkpoint type: {e}"),
        ))?;
```

Reply:
Lovely, thank you!